presto-main/src/main/java/io/prestosql/execution/scheduler/OptimizedLocalNodeSelector.java

```diff
@@ -0,0 +1,262 @@
+import io.prestosql.execution.NodeTaskMap;
+import io.prestosql.execution.RemoteTask;
+import io.prestosql.metadata.InternalNode;
+import io.prestosql.metadata.InternalNodeManager;
+import io.prestosql.metadata.Split;
+import io.prestosql.spi.HostAddress;
+import io.prestosql.spi.PrestoException;
+import org.apache.commons.math3.stat.descriptive.moment.StandardDeviation;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.prestosql.execution.scheduler.NodeScheduler.calculateLowWatermark;
+import static io.prestosql.execution.scheduler.NodeScheduler.randomizedNodes;
+import static io.prestosql.execution.scheduler.NodeScheduler.selectExactNodes;
+import static io.prestosql.execution.scheduler.NodeScheduler.selectNodes;
+import static io.prestosql.execution.scheduler.NodeScheduler.toWhenHasSplitQueueSpaceFuture;
+import static io.prestosql.spi.StandardErrorCode.NO_NODES_AVAILABLE;
+
+public class OptimizedLocalNodeSelector
+        extends SimpleNodeSelector
+{
+    private static final Logger log = Logger.get(OptimizedLocalNodeSelector.class);
+
+    OptimizedLocalNodeSelector(
+            InternalNodeManager nodeManager,
+            NodeTaskMap nodeTaskMap,
+            boolean includeCoordinator,
+            Supplier<NodeMap> nodeMap,
+            int minCandidates,
+            int maxSplitsPerNode,
+            int maxPendingSplitsPerTask)
+    {
+        super(nodeManager, nodeTaskMap, includeCoordinator, nodeMap, minCandidates, maxSplitsPerNode, maxPendingSplitsPerTask);
+    }
+
+    @Override
+    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)
+    {
+        Multimap<InternalNode, Split> assignment = HashMultimap.create();
+        NodeMap nodeMap = this.nodeMap.get().get();
+        NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
+
+        ResettableRandomizedIterator<InternalNode> randomCandidates = randomizedNodes(nodeMap, includeCoordinator, ImmutableSet.of());
+        Set<InternalNode> blockedExactNodes = new HashSet<>();
+        boolean splitWaitingForAnyNode = false;
+        // splitsToBeRedistributed remains false for internal splits or if force-local-scheduling is true
+        boolean splitsToBeRedistributed = false;
+
+        //prioritize assignment of splits to local nodes when the local nodes have slots available in the first pass
+        Set<Split> splitsToBeScheduled = new HashSet<>(splits);
+        for (Split split : splits) {
+            if (split.isRemotelyAccessible() && (split.getAddresses().size() > 0)
+                    && split.getAddresses().stream().anyMatch(address -> !address.getHostText().equals("localhost"))
+                    && !(CatalogName.isInternalSystemConnector(split.getCatalogName()) || split.getCatalogName().toString().equalsIgnoreCase("jmx"))) {
```

Comment:
Any specific reason why we check for information schema connector or jmx connector ? Plus it is not required as the splits created for these connectors are not remotely accessible so `split.isRemotelyAccessible()` itself might return false and exits the if statement

Reply:
I have removed that check as per your suggestion. It was relevant for our use case but is redundant here as you pointed out. Thanks for the observation.

Reply:
`split.getAddresses().stream().anyMatch(address -> !address.getHostText().equals("localhost"))`

Any specific reason for checking the addresses too. I guess this would be redundant too. As `selectExactNodes` might return an empty list if the address is `localHost`. (Not sure but I guess it should return an empty list)

Reply:
This check has been taken from https://github.com/prestosql/presto/blob/48336f218b950caee4beb381dde374d003f1176b/presto-hive/src/main/java/io/prestosql/plugin/hive/util/InternalHiveSplitFactory.java#L216
It is to prevent splits from entering the locality-assignment stage (and correspondingly the redistribution stage) when Presto is configured with data having no co-location with the worker nodes. In such a case, when data is read from s3, it has been seen that the splits have `localhost` by default in their address. This check makes sure that splits don't enter the first loop and the assignment is the same as the assignment made by SimpleNodeSelector.

Reply:
Oh okay but instead of doing this check. I think we can return an empty list if they have `localhost` by default in their address. (It is not related to the current PR then I think we don't need this check)